The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/25145925219/job/73705753151) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/73705753151) [(plain)](https://github.com/rust-lang/rust/commit/abc3a194770cd8c005becb53984060c617697e55/checks/73705753151/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
............F.............................         (142/142)

======== tests/rustdoc-gui/sidebar.goml ========

[ERROR] line 113: `.sidebar > .sidebar-crate > h2 > a` not found: for command `assert-text: (".sidebar > .sidebar-crate > h2 > a", "lib2")`
    at <file:///checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc-gui/doc/lib2/fn.foobar.html>


<= doc-ui tests done: 141 succeeded, 1 failed, 0 filtered out

Error: ()
```

</details>

--- a/compiler/rustc_middle/src/hir/map.rs
+++ b/compiler/rustc_middle/src/hir/map.rs
@@ -982,8 +982,8 @@ impl<'tcx> TyCtxt<'tcx> {
                 span,
                 ..
             }) => {
-                // Ensure that the returned span has the item's SyntaxContext.
-                fn_decl_span.find_ancestor_inside(*span).unwrap_or(*span)
+                // Ensure that the returned span has the closure expression's SyntaxContext.
+                fn_decl_span.find_ancestor_inside_same_ctxt(*span).unwrap_or(*span)
             }
             _ => self.hir_span_with_body(hir_id),
         };

--- a/tests/ui/proc-macro/auxiliary/closure-hir-span.rs
+++ b/tests/ui/proc-macro/auxiliary/closure-hir-span.rs
@@ -1,0 +1,23 @@
+//@ edition:2024
+
+extern crate proc_macro;
+
+#[proc_macro]
+pub fn m(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    let mut iter = input.into_iter();
+    let move_token = iter.next().unwrap();
+    let pipe1 = iter.next().unwrap();
+    let pipe2 = iter.next().unwrap();
+    let body = iter.next().unwrap();
+
+    let mut inner = proc_macro::TokenStream::new();
+    inner.extend([body]);
+    let new_body = proc_macro::TokenTree::Group(proc_macro::Group::new(
+        proc_macro::Delimiter::Brace,
+        inner,
+    ));
+
+    let mut out = proc_macro::TokenStream::new();
+    out.extend([move_token, pipe1, pipe2, new_body]);
+    out
+}

--- a/tests/ui/proc-macro/closure-hir-span.rs
+++ b/tests/ui/proc-macro/closure-hir-span.rs
@@ -1,0 +1,10 @@
+//! Regression test for <https://github.com/rust-lang/rust/issues/155724>
+//@ check-pass
+//@ edition:2024
+//@ proc-macro: closure-hir-span.rs
+
+extern crate closure_hir_span;
+
+fn main() {
+    closure_hir_span::m!(move || {});
+}

